newlib/libc/include/machine/_default_types.h

```diff
@@ -33,101 +33,206 @@
 #define __have_long32 1
 #endif
 
+/* Determine the width of integers if the compiler doesn't provide __X_WIDTH__ macros. */
+
+#if !defined(__SCHAR_WIDTH__)
+#if __EXP(SCHAR_MAX) == 0x7f
+#define __SCHAR_WIDTH__ 8
+#elif __EXP(SCHAR_MAX) == 0x7fff
+#define __SCHAR_WIDTH__ 16
+#elif __EXP(SCHAR_MAX) == 0x7fffffffL
+#define __SCHAR_WIDTH__ 32
+#endif
+#endif
+
+#if !defined(__SHRT_WIDTH__)
+#if __EXP(SHRT_MAX) == 0x7fff
+#define __SHRT_WIDTH__ 16
+#elif __EXP(SHRT_MAX) == 0x7fffffffL
+#define __SHRT_WIDTH__ 32
+#endif
+#endif
+
+#if !defined(__INT_WIDTH__)
+#if __EXP(INT_MAX) == 0x7fff
+#define __INT_WIDTH__ 16
+#elif __EXP(INT_MAX) == 0x7fffffffL
+#define __INT_WIDTH__ 32
+#elif __EXP(INT_MAX) > 0x7fffffffL
+#define __INT_WIDTH__ 64
+#endif
+#endif
+
+#if !defined(__LONG_WIDTH__)
+#if __have_long32
+#define __LONG_WIDTH__ 32
+#elif __have_long64
+#define __LONG_WIDTH__ 64
+#endif
+#endif
+
+#if !defined(__LONG_LONG_WIDTH__)
+#if __have_longlong64
+#define __LONG_LONG_WIDTH__ 64
+#endif
+#endif
+
+/* Determine the size of types if the compiler doesn't provide __SIZEOF_X__ macros. */
+
+#if !defined(__SIZEOF_SHORT__)
+#if __SHRT_WIDTH__ == 16
+#define __SIZEOF_SHORT__ 2
+#elif __SHRT_WIDTH__ == 32
+#define __SIZEOF_SHORT__ 4
+#endif
+#endif
+
+#if !defined(__SIZEOF_INT__)
+#if __INT_WIDTH__ == 16
+#define __SIZEOF_INT__ 2
+#elif __INT_WIDTH__ == 32
+#define __SIZEOF_INT__ 4
+#elif __INT_WIDTH__ == 64
+#define __SIZEOF_INT__ 8
+#endif
+#endif
+
+#ifndef __SIZEOF_LONG__
+#if __LONG_WIDTH__ == 32
+#define __SIZEOF_LONG__ 4
+#elif __LONG_WIDTH__ == 64
+#define __SIZEOF_LONG__ 8
+#endif
+#endif
+
+#if (!defined(__SIZEOF_LONG_LONG__) && __LONG_LONG_WIDTH__ == 64)
+#define __SIZEOF_LONG_LONG__ 8
+#endif
+
+#if (!defined(__SIZEOF_FLOAT__) && defined(__FLT_MANT_DIG__))
+#if __FLT_MANT_DIG__ == 24
+#define __SIZEOF_FLOAT__ 4
+#else
+#error "unexpected __FLT_MANT_DIG__ value"
+#endif
+#endif
+
+#if (!defined(__SIZEOF_DOUBLE__) && defined(__DBL_MANT_DIG__))
+#if __DBL_MANT_DIG__ == 53
+#define __SIZEOF_DOUBLE__ 8
+#else
+#error "unexpected __DBL_MANT_DIG__ value"
+#endif
+#endif
+
+#if (!defined(__SIZEOF_LONG_DOUBLE__) && defined(__LDBL_MANT_DIG__))
+#if __LDBL_MANT_DIG__ == 53
+#define __SIZEOF_LONG_DOUBLE__ 8
+#elif __LDBL_MANT_DIG__ == 64 || __LDBL_MANT_DIG__ == 113
+#define __SIZEOF_LONG_DOUBLE__ 16
+#else
+#error "unexpected __LDBL_MANT_DIG__ values"
+#endif
+#endif
+
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+/* Select type of fixed width integers if the compiler doesn't specify them. */
+
+#if (!defined(__INT8_TYPE__) && __SCHAR_WIDTH__ == 8)
+#define __INT8_TYPE__ char
+#endif
+
+#if !defined(__UINT8_TYPE__) && defined(__INT8_TYPE__)
+#define __UINT8_TYPE__ unsigned __INT8_TYPE__
```

Comment:
this can probably just be 'unsigned char' so we can use 'signed char' above.

Reply:
:+1: